TST use global_random_seed in sklearn/mixture/tests/test_gaussian_mixture.py

OmarManzoor opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #22827

#### What does this implement/fix? Explain your changes.
- Added global_random_seed in applicable tests of test_gaussian_mixture.

#### Any other comments?
- Some tests were only tested for 10 seeds for the range 0-9 since all 100 seeds take too long.
- For some of the tests the training samples or features were increased to ensure that the tests pass for the different seeds.
- Some tests are already using seed through parametrization. Do we need to use global_random_seed with them? Currently I left them as they are.
- test_gaussian_mixture_fit takes quite long even with only 10 seeds so I left it.
- test_gaussian_mixture_fit_best_params takes about 0.95 s with 10 seeds and gives two errors. So if I try to increase the training set size or change parameters it goes beyond 1s so I did not modify it.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

OmarManzoor committed:
TST use global_random_seed in sklearn/mixture/tests/test_gaussian_mixture.py

OmarManzoor committed:
Addition of global random seed in test_gaussian_mixture_fit_predict_n_init

OmarManzoor committed:
Add global_random_seed to test_multiple_init

OmarManzoor committed:
PR changes

OmarManzoor commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -478,9 +478,14 @@ def test_gaussian_mixture_log_probabilities():
+    # be too costly. Restricting to the 0-9 range is necessary to
+    # use small enough datasets that avoid increasing the run time
+    # too much.
+    rng = np.random.RandomState(global_random_seed % 10)
```
I have adjusted the PR. However just to clarify there are certain tests which take a while to run for example:
- test_multiple_init - 4.27s
- test_gaussian_mixture_fit_predict_n_init - 6.64s
- test_property - 3.04s

Also I tried to experiment with various parameters to generate the random data, but I was not able to make the test_gaussian_mixture_fit pass for all seeds.

jeremiedbb replied:
The global_random_seed does not change the duration of the tests. We always run for a single seed. It's just that we don't always run with the same seed.

betatim commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -634,9 +639,9 @@ def test_gaussian_mixture_fit():
-    rand_data = RandomData(rng)
+def test_gaussian_mixture_fit_best_params(global_random_seed):
+    rng = np.random.RandomState(global_random_seed)
+    rand_data = RandomData(rng, n_samples=400, n_features=3, scale=100)
```
```suggestion
    rand_data = RandomData(rng)
```
There should be no need to change this.

betatim replied:
I'm not super familiar with the code that is being tested here, it looks like the test compares performing ten individual fits to the result of doing one fit with ten "random inits".

So naively I'd expect that this test should either always pass or always fail. Independent of which random seed is chosen. Do you have an idea why changing the dataset in the way you did helps? From the difference in the results it looks like they are almost but not quite the same. So maybe increasing the size of the dataset will help with statistical fluctuations that lead to the difference? But that requires a bit of knowledge of the algorithm I think.

OmarManzoor replied:
I agree with you that to actually understand the behavior or why the test passes with a small increase in the dataset, I would have to dig deeper into the actual code and working behind the EM algorithm.

OmarManzoor replied:
@betatim I did check this a bit and it seems the randomness is involved in a combination of the data generation and in computing the centroids in Kmeans. So it is possible some random seeds produce a combination which produces slightly different results overall when we compare the individual 10 fits with the single fit with 10 initializations. Therefore for now I just removed the global_random_seed from these two tests.

betatim replied:
Thanks for digging into this. If we can't make the call to the random number generators exactly the same, so that they get the same sequence of random numbers then we have to allow from some fluctuations. Like you said.

One way out is to increase the dataset, in the hope of reducing the size of the fluctuations.

I think the reason it works with the current setup (what is in `main`), must be pure luck then, right?

I wonder if there is a way to make the test's assert statement more forgiving, basically "the two numbers are equal to within the statistical precision we have" or something like that.

OmarManzoor replied:
I think we can use assert_allclose along with adjusting atol or rtol. However we can also increase the dataset like I had done originally and that passes all the seeds.

jeremiedbb replied:
The random_state was not used correctly. To be comparable, the rng has to be consumed the same way in both cases. I fixed the test and now it passes

jeremiedbb replied:
Also, the test should not have passed in the first place because it was wrong. Since we compare scores, we need to take the max because for a score bigger is better.

OmarManzoor committed:
Adjustment in test_gaussian_mixture_aic_bic

OmarManzoor committed:
Merge remote-tracking branch 'upstream/main' into global_random_seed_gaussian_mixture

OmarManzoor committed:
Remove global_random_seed from two tests

OmarManzoor committed:
Increase the dataset for test_gaussian_mixture_fit_best_params

jeremiedbb committed:
global_random_seed for more tests

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/OmarManzoor/24616

jeremiedbb committed:
enable more tests + tweaks

jeremiedbb committed:
[all random seeds]
test_suffstat_sk_full
test_suffstat_sk_tied
test_suffstat_sk_diag
test_gaussian_suffstat_sk_spherical
test_compute_log_det_cholesky
test_gaussian_mixture_log_probabilities
test_gaussian_mixture_estimate_log_prob_resp
test_gaussian_mixture_predict_predict_proba
test_gaussian_mixture_fit_predict
test_gaussian_mixture_fit_predict_n_init
test_gaussian_mixture_fit_best_params
test_multiple_init
test_bic_1d_1component
test_gaussian_mixture_aic_bic
test_warm_start
test_convergence_detected_with_warm_start
test_score
test_monotonic_likelihood
test_regularisation
test_property
test_sample
test_init
test_init_means_not_duplicated
test_means_for_all_inits
test_gaussian_mixture_precisions_init_diag

jeremiedbb committed:
chose best run based on score

jeremiedbb committed:
Merge branch 'fix-mixture-lower-bound' into pr/OmarManzoor/24616

jeremiedbb committed:
fix n_init

jeremiedbb committed:
[all random seeds]
test_suffstat_sk_full
test_suffstat_sk_tied
test_suffstat_sk_diag
test_gaussian_suffstat_sk_spherical
test_compute_log_det_cholesky
test_gaussian_mixture_log_probabilities
test_gaussian_mixture_estimate_log_prob_resp
test_gaussian_mixture_predict_predict_proba
test_gaussian_mixture_fit_predict
test_gaussian_mixture_fit_predict_n_init
test_gaussian_mixture_fit_best_params
test_multiple_init
test_bic_1d_1component
test_gaussian_mixture_aic_bic
test_warm_start
test_convergence_detected_with_warm_start
test_score
test_monotonic_likelihood
test_regularisation
test_property
test_sample
test_init
test_init_means_not_duplicated
test_means_for_all_inits
test_gaussian_mixture_precisions_init_diag

jeremiedbb committed:
[azure parallel] [all random seeds]
test_suffstat_sk_full
test_suffstat_sk_tied
test_suffstat_sk_diag
test_gaussian_suffstat_sk_spherical
test_compute_log_det_cholesky
test_gaussian_mixture_log_probabilities
test_gaussian_mixture_estimate_log_prob_resp
test_gaussian_mixture_predict_predict_proba
test_gaussian_mixture_fit_predict
test_gaussian_mixture_fit_predict_n_init
test_gaussian_mixture_fit_best_params
test_multiple_init
test_bic_1d_1component
test_gaussian_mixture_aic_bic
test_warm_start
test_convergence_detected_with_warm_start
test_score
test_monotonic_likelihood
test_regularisation
test_property
test_sample
test_init
test_init_means_not_duplicated
test_means_for_all_inits
test_gaussian_mixture_precisions_init_diag

jeremiedbb commented:
I think the way the best model was selected when running multiple inits was wrong. 
- First, it was based on the ``lower_bound`` instead of the ``log_prob_norm`` (which is the same thing that the ``score`` computes). I think we want to pick the model that gives the best score. For ``GaussianMixture`` it's the same but not for ``BayesianGaussianMixture``.
- Then the final E-step should be done at the end of each run to be sure we compare the models based of the score they have, not based on the score before the last E-step.

This fix does not require a new non-regression test imo. The non-regressions test are the ones already present in the test file. They now pass for all random seeds.

I'd like to have the opinion of someone familiar with the mixture models though. Maybe @ogrisel ?

OmarManzoor commented:
@ogrisel @jeremiedbb What should we do about this PR?

OmarManzoor commented:
Closing, if required can be reopened.
